File: bindings/php/src/Exception/DatalogicException.php
<?php

declare(strict_types=1);

/* SPDX-License-Identifier: Apache-2.0 */

namespace Goplasmatic\Datalogic\Exception;

use FFI;
use FFI\CData;
use Goplasmatic\Datalogic\Internal\Native;
use RuntimeException;

/**
 * Base class for every exception thrown by this binding. Carries the
 * structured error detail exposed by the C ABI's error handles
 * (`datalogic_error_message` / `_tag` / `_operator` / `_path_json`).
 */
class DatalogicException extends RuntimeException
{
    public function __construct(
        string $message,
        public readonly ?string $errorType = null,
        public readonly ?string $operatorName = null,
        public readonly ?string $pathJson = null,
    ) {
        parent::__construct($message);
    }

    /**
     * Construct the right subclass from a failed v2 call: `$status` is
     * the `datalogic_status` the call returned, `$errOut` the
     * `datalogic_error *` out-param slot it may have filled. Reads the
     * borrowed accessors and ALWAYS releases the handle.
     *
     * @internal
     */
    public static function fromNative(int $status, CData $errOut, string $fallback): self
    {
        $ffi = Native::ffi();
        if (FFI::isNull($errOut)) {
            // No detail captured (should not happen — every fallible
            // call in this binding passes an out-param).
            return $status === Native::STATUS_PARSE
                ? new ParseException($fallback)
                : new EvaluateException($fallback);
        }

        $len = $ffi->new('size_t');
        $addr = FFI::addr($len);
        $msg  = Native::copyBytes($ffi->datalogic_error_message($errOut, $addr), $len->cdata);
        $tag  = Native::copyBytes($ffi->datalogic_error_tag($errOut, $addr), $len->cdata);
        $op   = Native::copyBytes($ffi->datalogic_error_operator($errOut, $addr), $len->cdata);
        $path = Native::copyBytes($ffi->datalogic_error_path_json($errOut, $addr), $len->cdata);
        $ffi->datalogic_error_free($errOut);

        $message = ($msg === null || $msg === '') ? $fallback : $msg;
        return $status === Native::STATUS_PARSE
            ? new ParseException($message, $tag, $op, $path)
            : new EvaluateException($message, $tag, $op, $path);
    }
}

File: bindings/dotnet/src/Datalogic/DatalogicException.cs
// SPDX-License-Identifier: Apache-2.0

using System.Text;

using Goplasmatic.Datalogic.Native;

namespace Goplasmatic.Datalogic;

/// <summary>
/// Base class for every exception thrown by this binding. Carries the
/// structured error detail exposed by the C ABI's error handles
/// (`datalogic_error_*` accessors).
/// </summary>
public class DatalogicException : Exception
{
    /// <summary>Stable error tag from the engine (e.g. "ParseError", "Thrown", "NaN", "TypeMismatch").</summary>
    public string? ErrorType { get; }
    /// <summary>Outermost failing operator name (e.g. "+"), or null if not operator-scoped.</summary>
    public string? Operator { get; }
    /// <summary>Resolved root-to-leaf error path as a JSON array, or null if not available.</summary>
    public string? PathJson { get; }
    /// <summary>Coarse status the failing native call returned.</summary>
    public EvaluationStatus Status { get; }

    internal DatalogicException(
        string message,
        string? errorType,
        string? @operator,
        string? pathJson,
        EvaluationStatus status = EvaluationStatus.EvaluationError)
        : base(message)
    {
        ErrorType = errorType;
        Operator = @operator;
        PathJson = pathJson;
        Status = status;
    }

    /// <summary>
    /// Consume a `datalogic_error *` handle produced by a failing native
    /// call: read the accessors, ALWAYS free the handle, and construct
    /// the mapped exception subclass. <paramref name="err"/> may be
    /// <see cref="IntPtr.Zero"/> (no capture) — then the
    /// <paramref name="fallback"/> message and the raw status drive the
    /// mapping.
    /// </summary>
    internal static DatalogicException FromNative(DatalogicStatus status, IntPtr err, string fallback)
    {
        var message = fallback;
        string? tag = null, op = null, path = null;
        if (err != IntPtr.Zero)
        {
            try
            {
                // The handle's own status is authoritative (identical to
                // the returned one by construction).
                status = NativeMethods.datalogic_error_status(err);
                unsafe
                {
                    message = Read(NativeMethods.datalogic_error_message(err, out var len), len) ?? fallback;
                    tag = Read(NativeMethods.datalogic_error_tag(err, out len), len);
                    op = Read(NativeMethods.datalogic_error_operator(err, out len), len);
                    path = Read(NativeMethods.datalogic_error_path_json(err, out len), len);
                }
            }
            finally
            {
                NativeMethods.datalogic_error_free(err);
            }
        }
        return Create((EvaluationStatus)status, tag, message, op, path);
    }

    /// <summary>
    /// Map (status, tag) to the public exception subclass: parse
    /// failures become <see cref="ParseException"/>, everything else
    /// <see cref="EvaluateException"/> — the same split the v1 binding
    /// exposed.
    /// </summary>
    internal static DatalogicException Create(
        EvaluationStatus status,
        string? tag,
        string message,
        string? @operator,
        string? pathJson)
        => status == EvaluationStatus.ParseError || tag == "ParseError"
            ? new ParseException(message, tag, @operator, pathJson, status)
            : new EvaluateException(message, tag, @operator, pathJson, status);

    private static unsafe string? Read(byte* ptr, nuint len)
        => ptr == null ? null : Encoding.UTF8.GetString(ptr, checked((int)len));
}

/// <summary>Thrown when a JSONLogic rule (or data / config JSON) fails to parse.</summary>
public sealed class ParseException : DatalogicException
{
    internal ParseException(
        string message,
        string? errorType,
        string? @operator,
        string? pathJson,
        EvaluationStatus status = EvaluationStatus.ParseError)
        : base(message, errorType, @operator, pathJson, status) { }
}

/// <summary>Thrown when rule evaluation fails (runtime error, thrown, NaN, type mismatch, …).</summary>
public sealed class EvaluateException : DatalogicException
{
    internal EvaluateException(
        string message,
        string? errorType,
        string? @operator,
        string? pathJson,
        EvaluationStatus status = EvaluationStatus.EvaluationError)
        : base(message, errorType, @operator, pathJson, status) { }
}
